internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -336,7 +348,53 @@ func (r *BMACReconciler) handleBMHFinalizer(ctx context.Context, log logrus.Fiel
 		return reconcileComplete{stop: true, dirty: true}
 	}
 
-	if dirty := r.configureBMHForCleaning(ctx, log, bmh); dirty {
+	drainTimedOut, err := nodeDrainTimedout(bmh)
+	if err != nil {
+		message := fmt.Sprintf("failed to check node drain timeout: %s", err.Error())
+		log.Error(message)
+		setAnnotation(&bmh.ObjectMeta, BMH_NODE_DRAIN_STATUS_ANNOTATION, message)
+		return reconcileComplete{stop: true, dirty: true}
+	}
+
+	if !drainTimedOut && bmh.GetAnnotations()[BMH_NODE_DRAIN_STATUS_ANNOTATION] != drainStatusSuccess {
+		dirty := false
+		log.Info("Draining agent node...")
+		if _, present := bmh.GetAnnotations()[BMH_NODE_DRAIN_START_ANNOTATION]; !present {
+			setAnnotation(&bmh.ObjectMeta, BMH_NODE_DRAIN_START_ANNOTATION, time.Now().UTC().Truncate(time.Second).Format(time.RFC3339))
+			dirty = true
+
+		}
+		requeue, err := r.drainAgentNode(ctx, log, agent)
```

Comment:
A single call will take no longer than 20 seconds.
We set a timeout [here](https://github.com/openshift/assisted-service/pull/5110/files#diff-86415614c6ed40c56c5ca2b5f751a65746ef5c83b61e6f021a4f5d817e4425baR1682)

I don't want to use goroutines because it makes this significantly more difficult to track.
Each reconcile loop would have to know if there was a drain routine already running and if to start a new one for this.

I took this logic (including the timeout values and behavior) directly from the CAPI controller implementation so I'm hoping it will work fine for us if it works for them.

Reply:
Do we have the controller set for multiple concurrent reconciles at least? The default is `1`, so we should make sure we've configured it to have additional available, so other resources can be reconciled, especially in case there are multiple drains in a row.

Reply:
It gets set here: https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.6/pkg/controller#Options

Reply:
Or one thread might be waiting on drain while another sees that the drain has timed out and deletes the BMH.

I think the worst that will happen is that we get an error about the object already being deleted or already being modified.

Reply:
Because bmh controller doesn't access the DB then there should not be any races with the DB. 
If this configuration is per controller then we have a lower risk enabling it for the bmh controller, after all the controller know how to handle object versioning, the issue can be that we will not handle it well in the code, meaning that you will not be able to submit something because the object have changed by another reconcile and you will just skip or log an error. 

BMH is handling very specific cases, if there will be delay and some host will not boot or will not update the url on time it's not a big deal. 

Because it can cause unexpected behavior and we are close to the feature freeze, i suggest to work with a the default `MaxConcurrentReconciles = 1` and open a different ticket to handle  it in 2.9.0

Reply:
I can also pretty easily add config for this with a default to 1 if we want

Reply:
Added the option for a user to set the number of concurrent reconciles. Currently set the default to 1 since this entire feature is opt-in.

Reply:
Is the expectation then that if a user wants to use this feature, they should also adjust the number of concurrent reconciles? How will they know to do that?

If we'll have some users running with concurrent reconciles, it would help to get as much testing of that configuration as we can. Making multiple the default would ensure it gets maximum testing, and then we have an easy way to go back to 1.

If we're not confident in being able to run concurrent reconciles with stability, I'm not sure how we can confidently deliver this feature as a good experience, unless we go with the other approach of doing drains in dedicated goroutines.

Thoughts?

Reply:
I'm not really concerned.

I don't think the delay in processing reconciles when a node is getting drained is going to impact user experience in a very noticeable way.
Also I think the people using this feature are a significant minority. Adding concurrency for the few people using this at the risk of causing difficult to debug issues for the majority of users isn't a good idea this late in the release cycle IMO.

I'd be okay enabling it by default after this PR gets backported.